### ISSUE TITLE
Ensure we detect complete replacement of etcd cluster

### ISSUE THREAD

**Issue by fasaxc** [Metaswitch/calico#340]
If the etcd server is replaced wholesale with another one running on the same IP address, felix could issue a poll to the new server for an etcd_index that is far in the future as far as that new server is concerned.

This would be hard to hit in a traditional mainline etcd deployment but some people may run etcd on a RAM disk as an ephemeral store with a floating IP, making it much easier to hit this window.

etcd clusters do have UUIDs so if there's some way we can check the UUID hasn't changed since our last poll, that'd do the trick.

**Comment by fasaxc** [comment]
Looks like there's at least a partial solution.  When etcd responds, the response includes the cluster ID.  There's just the slight wrinkle that, if a poll fails, we don't get a response.

**Comment by fasaxc** [comment]
I asked the etcd devs if there's a way to close that window already: https://github.com/coreos/etcd/issues/2649

**Comment by fasaxc** [comment]
Turns out that etcd writes its headers at the start of the long poll, including the one we need.  We'll need to get access to the header but it's already doing what we need.

They also seem to support streaming events via an option: stream=true.  I couldn't find reference to that in the docs but, if we could use it then that'd prevent us from needing to reconnect every poll as we do now.

**Comment by fasaxc** [comment]
This urllib3 incantation does the trick to load from their streaming model:
```
>>> import urllib3
>>> http = urllib3.HTTPConnectionPool("localhost",4001)
>>> r = http.request_encode_url("GET","/v2/keys/calico/",{"wait":"true", "stream":"true", "recursive":"true"}, preload_content=False)
>>> r.getheaders()
HTTPHeaderDict({'transfer-encoding': 'chunked', 'x-raft-index': '661276', 'x-raft-term': '7', 'x-etcd-index': '116', 'date': 'Fri, 10 Apr 2015 05:10:23 GMT', 'cont
ent-type': 'application/json', 'x-etcd-cluster-id': '7e27652122e8b2ae'})
>>> r.readline()
'{"action":"set","node":{"key":"/calico/foo","value":"baz","modifiedIndex":117,"createdIndex":117},"prevNode":{"key":"/calico/foo","value":"baz","modifiedIndex":11
6,"createdIndex":116}}\n'
```
I'm wondering if we should just drop python-etcd; it's a long way from supporting this.

**Comment by Lukasa** [comment]
python-etcd is such a tiny wrapper it's really hard to see what it gets us. The other option, though, is to fork it and add the feature we need ourselves.

**Comment by fasaxc** [comment]
Fixed in #384
